src/main/java/io/numaproj/numaflow/sinker/DatumIterator.java

```diff
@@ -0,0 +1,26 @@
+package io.numaproj.numaflow.sinker;
+
+/**
+ * An iterator over a collection of {@link Datum} elements.
+ * Passed to {@link Sinker#processMessages(DatumIterator)} method.
+ */
+public interface DatumIterator {
+    /**
+     * Returns true if there is at least one more element in the iterator, or false otherwise.
+     * This method blocks until an element becomes available in the blocking queue.
+     * This method should be called before calling {@link #next()} to ensure that the iterator is not empty.
+     * otherwise {@link #next()} will throw an {@link IllegalStateException}.
+     *
+     * @return true if there is at least one more element in the iterator, or false otherwise
+     */
+    boolean hasNext();
+
+    /**
+     * Returns the next element in the iterator
+     *
+     * @return the next element in the iterator
+     *
+     * @throws IllegalStateException if there are no more elements in the iterator
+     */
+    Datum next();
```

Comment:
Can we combine these two into one?

```
// Return null if the iterator doesn't have a next
Datum next();
```

Reply:
https://docs.oracle.com/javase/8/docs/api/java/util/Iterator.html, java Iterator follows this pattern

Reply:
Agreed. Having `hasNext()` makes it more flexible for users to operate on the iterator.